### DISCUSSION TITLE
ModuleNotFoundError from all environments after installing XGBoost

### DISCUSSION THREAD

**Discussion by U0M0Z** [rdkit/rdkit#3771] [Q&A]
Hi everyone, 
I have been using RDkit for the last couple of months in combination with mol2vec (installed via pip install https://github.com/...)

I am an intermediate python and conda user and I usually work developing conda environments for the project I am working on. I have come across a problem that I cannot solve, I have installed the xgboost library for some tasks (using pip install xgboost), since that moment I cannot use rdkit. Notably I cannot import rdkit both from the terminal and in jupyter notebooks in any of the environments even if present. This is error in both jupyter notebooks and in the terminal:

>>> import rdkit
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'rdkit'

Or:

>>> from rdkit import Chem
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'rdkit'

If I check the environment with conda list I have both the xgboost (installed via xgboost and py-xgboost) and rdkit installed:

####
#### Name                                        Version                                       Build  Channel
>>> _py-xgboost-mutex                             2.0                       cpu_0    conda-forge
appnope                                       0.1.2           py37hecd8cb5_1001  
argon2-cffi                                   20.1.0           py37haf1e3a3_1  
async_generator                               1.10             py37h28b3542_0  
attrs                                         20.3.0             pyhd3eb1b0_0  
backcall                                      0.2.0                      py_0  
blas                                          1.0                         mkl  
bleach                                        3.2.1                      py_0  
boost                                         1.68.0          py37h9888f84_1001    conda-forge
boost-cpp                                     1.68.0            h6f8c590_1000    conda-forge
boto                                          2.49.0                   py37_0    anaconda
boto3                                         1.16.0                     py_0    anaconda
botocore                                      1.19.52            pyhd8ed1ab_0    conda-forge
brotlipy                                      0.7.0           py37h9ed2024_1003  
bz2file                                       0.98                     pypi_0    pypi
bzip2                                         1.0.8                h1de35cc_0  
ca-certificates                               2020.10.14                    0    anaconda
cachetools                                    4.2.0              pyhd3eb1b0_0  
cairo                                         1.14.12              hc4e6be7_4  
certifi                                       2020.6.20                py37_0    anaconda
cffi                                          1.14.4           py37h2125817_0  
chardet                                       4.0.0           py37hecd8cb5_1003  
cryptography                                  3.3.1            py37hbcfaee0_0  
cycler                                        0.10.0                   py37_0  
decorator                                     4.4.2                      py_0  
defusedxml                                    0.6.0                      py_0  
entrypoints                                   0.3                      py37_0  
fontconfig                                    2.13.0               h5d5b041_1  
freetype                                      2.10.4               ha233b18_0  
gensim                                        3.8.0            py37h6440ff4_0    anaconda
gettext                                       0.19.8.1             hb0f4f8b_2  
glib                                          2.66.1               h9bbe63b_0  
google-api-core                               1.22.2                   py37_0  
google-auth                                   1.21.3                     py_0  
google-cloud-core                             1.5.0              pyhd3eb1b0_0  
google-cloud-storage                          1.35.0             pyhd3eb1b0_0  
google-crc32c                                 1.1.0            py37h9ed2024_1  
google-resumable-media                        1.2.0              pyhd3eb1b0_1  
googleapis-common-protos                      1.52.0                   py37_0  
icu                                           58.2                 h0a44026_3  
idna                                          2.10                       py_0  
importlib-metadata                            2.0.0                      py_1  
importlib_metadata                            2.0.0                         1  
intel-openmp                                  2019.4                      233  
ipykernel                                     5.3.4            py37h5ca1d4c_0  
ipython                                       7.19.0           py37h01d92e1_0  
ipython_genutils                              0.2.0              pyhd3eb1b0_1  
jedi                                          0.18.0           py37hecd8cb5_0  
jinja2                                        2.11.2                     py_0  
jmespath                                      0.10.0                     py_0    anaconda
joblib                                        1.0.0              pyhd3eb1b0_0  
jpeg                                          9b                   he5867d9_2  
json5                                         0.9.5                      py_0  
jsonschema                                    3.2.0                      py_2  
jupyter_client                                6.1.7                      py_0  
jupyter_core                                  4.7.0            py37hecd8cb5_0  
jupyterlab                                    2.2.6                      py_0  
jupyterlab_pygments                           0.1.2                      py_0  
jupyterlab_server                             1.2.0                      py_0  
kiwisolver                                    1.3.0            py37h23ab428_0  
lcms2                                         2.11                 h92f6f08_0  
libboost                                      1.73.0              ha68ee2f_10  
libcrc32c                                     1.1.1                hb1e8313_2  
libcxx                                        10.0.0                        1  
libedit                                       3.1.20191231         h1de35cc_1  
libffi                                        3.3                  hb1e8313_2  
libgfortran                                   3.0.1                h93005f0_2  
libiconv                                      1.16                 h1de35cc_0  
libpng                                        1.6.37               ha441bb4_0  
libprotobuf                                   3.13.0.1             hab81aa3_0  
libsodium                                     1.0.18               h1de35cc_0  
libtiff                                       4.1.0                hcb84e12_1  
libxgboost                                    1.1.1                h4a8c4bd_0    conda-forge
libxml2                                       2.9.10               h7cdb67c_3  
llvm-openmp                                   10.0.0               h28b9765_0  
lz4-c                                         1.9.2                h79c402e_3  
markupsafe                                    1.1.1            py37h1de35cc_0  
matplotlib                                    3.3.2                hecd8cb5_0  
matplotlib-base                               3.3.2            py37h181983e_0  
mistune                                       0.8.4            py37h1de35cc_0  
mkl                                           2019.4                      233  
mkl-service                                   2.3.0            py37h9ed2024_0  
mkl_fft                                       1.2.0            py37hc64f4ea_0  
mkl_random                                    1.1.1            py37h959d312_0  
mol2vec                                       0.1                      pypi_0    pypi
multidict                                     5.1.0            py37h9ed2024_2  
nbclient                                      0.5.1                      py_0  
nbconvert                                     6.0.7                    py37_0  
nbformat                                      5.0.8                      py_0  
ncurses                                       6.2                  h0a44026_1  
nest-asyncio                                  1.4.3              pyhd3eb1b0_0  
notebook                                      6.1.6            py37hecd8cb5_0  
numpy                                         1.19.2           py37h456fd55_0  
numpy-base                                    1.19.2           py37hcfb5961_0  
olefile                                       0.46                     py37_0  
openssl                                       1.1.1i               h9ed2024_0  
packaging                                     20.8               pyhd3eb1b0_0  
pandas                                        1.2.0            py37hb2f4e1b_0  
pandoc                                        2.11                 h0dc7051_0  
pandocfilters                                 1.4.3            py37hecd8cb5_1  
parso                                         0.7.0                      py_0  
pcre                                          8.44                 hb1e8313_0  
pexpect                                       4.8.0              pyhd3eb1b0_3  
pickleshare                                   0.7.5           pyhd3eb1b0_1003  
pillow                                        8.1.0            py37h5270095_0  
pip                                           20.3.3           py37hecd8cb5_0  
pixman                                        0.40.0               haf1e3a3_0  
prometheus_client                             0.9.0              pyhd3eb1b0_0  
prompt-toolkit                                3.0.8                      py_0  
protobuf                                      3.13.0.1         py37hb1e8313_1  
ptyprocess                                    0.7.0              pyhd3eb1b0_2  
py-boost                                      1.73.0          py37he9a2976_10  
py-xgboost                                    1.1.1            py37hc8dfbb8_0    conda-forge
pyasn1                                        0.4.8                      py_0  
pyasn1-modules                                0.2.8                      py_0  
pycparser                                     2.20                       py_2  
pygments                                      2.7.3              pyhd3eb1b0_0  
pyopenssl                                     20.0.1             pyhd3eb1b0_1  
pyparsing                                     2.4.7                      py_0  
pyrsistent                                    0.17.3           py37haf1e3a3_0  
pysocks                                       1.7.1            py37hecd8cb5_0  
python                                        3.7.9                h26836e1_0  
python-dateutil                               2.8.1                      py_0  
python_abi                                    3.7                     1_cp37m    conda-forge
pytz                                          2020.5             pyhd3eb1b0_0  
pyzmq                                         20.0.0           py37h23ab428_1  
rdkit                                         2020.09.1.0      py37h81f2deb_1    rdkit
readline                                      8.0                  h1de35cc_0  
requests                                      2.25.1             pyhd3eb1b0_0  
rsa                                           4.7                pyhd3eb1b0_0  
s3transfer                                    0.3.3                    py37_0    anaconda
scikit-learn                                  0.23.2           py37h959d312_0  
scipy                                         1.5.2            py37h912ce22_0  
seaborn                                       0.11.1             pyhd3eb1b0_0  
send2trash                                    1.5.0              pyhd3eb1b0_1  
setuptools                                    51.1.2           py37hecd8cb5_3  
six                                           1.15.0           py37hecd8cb5_0  
smart_open                                    2.0.0                      py_0  
sqlite                                        3.33.0               hffcf06c_0  
terminado                                     0.9.2            py37hecd8cb5_0  
testpath                                      0.4.4                      py_0  
threadpoolctl                                 2.1.0              pyh5ca1d4c_0  
tk                                            8.6.10               hb0a8c7a_0  
tornado                                       6.1              py37h9ed2024_0  
tqdm                                          4.55.1             pyhd3eb1b0_0  
traitlets                                     5.0.5                      py_0  
urllib3                                       1.26.2             pyhd3eb1b0_0  
wcwidth                                       0.2.5                      py_0  
webencodings                                  0.5.1                    py37_1  
wheel                                         0.36.2             pyhd3eb1b0_0  
xgboost                                       1.1.1            py37h570ac47_0    conda-forge
xz                                            5.2.5                h1de35cc_0  
yarl                                          1.5.1            py37haf1e3a3_0  
zeromq                                        4.3.3                hb1e8313_3  
zipp                                          3.4.0              pyhd3eb1b0_0  
zlib                                          1.2.11               h1de35cc_3  
zstd                                          1.4.5                h41d2c2f_0  
>>>

Python version is 3.7.3

Everything was working normally before installing the xgboost, besides having to reinstall RDKit millions of times due to problems with C++ signatures and boost, boost-cpp. 

This time I have tried to reinstall and recreate the environment from scratch (without py-xgboost and xgboost, but also without mol2vec). I have tried to install through:

conda create -c rdkit -n my-rdkit-env rdkit

via 

conda install conda-forge::rdkit

I have tried to use python 3.7, 3.8, 3.9 and still it is not working in any scenario: I cannot import RDKit from any of the environments even with the most simple one. I can still use other libraries without any problem, like pandas, xgboost, numpy, sciikit-learn, etc.

Any idea/suggestion would be much appreciated because I cannot solve this problem alone. Is it a python related problem? 

Thanks, 

Here you have few specs

MacOS Catalina 
Version 10.15.7

(XGBoost_RDKit2021) mac-slibekki:bin tommaso$ which python
/Users/tommaso/anaconda3/bin/python

(XGBoost_RDKit2021) mac-slibekki:bin tommaso$ python -V
Python 3.7.3

(XGBoost_RDKit2021) mac-slibekki:bin tommaso$ conda info --base
/Users/tommaso/anaconda3

**Comment by greglandrum** [comment]
@U0M0Z apologies that getting these complex environments working is so tricky.
Unfortunately diagnosing the problems is also hard.

To try and build up systematically, I just created an environment on my Mac with the RDKit and xgboost and everything works fine. Here are the commands for that, along with small tests to make sure things are working. Note that, since I'm impatient, I'm using mamba instead of conda to do installs and that I install the RDKit from conda-forge instead of the rdkit channel... 
```
conda create -n py_rdkit_xgb python=3.8
conda activate py_rdkit_xgb
conda install -c conda-forge mamba
mamba install -c conda-forge rdkit xgboost
python -c 'import rdkit;print(rdkit.__file__)'
```
Everything works fine here.

As an aside: I installed xgboost from conda-forge as your environment indicates.

Next step, installing mol2vec via:
```
pip install git+https://github.com/samoturk/mol2vec
```
That also worked, as did running the RDKit afterwards.

So it looks like there's no inherent problem with compatibility of these three packages. I guess what has likely happened is that your environment has somehow ended up corrupted.

Can you please try the simple experimental environment I show here and see if that works for you? You'll likely also need to install things like jupyter in it as well... if you can I would suggest installing this (and anything else you can) from conda-forge.

**Comment by greglandrum** [comment]
Marking this as "question" and not "bug" and moving it to discussions because this is not an RDKit bug.

**Reply by U0M0Z** [reply]
Hi Greg, thanks for the prompt answer and sorry for the late reply. 

The solution you have suggested seems to work and I can use both libraries in the created environment. Besides, I have noticed my old environments started working again (somehow). 

I just wanted to add something: I have noticed that if I try to install py-xgboost in the new environment, there are a lot of conflicts within the packages. Maybe it was something related to the co-presence of boost, boost-cpp and py-xgboost. I'll try to give you an update if I notice the error again.

Thanks again

**Reply by greglandrum** [reply]
@U0M0Z it's somewhat odd that you are seeing conflicts with `py-xgboost`. If I follow the steps listed above, I end up with py-xgboost installed without any problems (it comes automatically with xgboost from conda-forge).

**Comment by kienerj** [comment]
Complex environments are tricky especially if you need to update them. usually just simpler to remake from scratch. Since I played with this not too long ago (on Windows) and had issues as well, below the yml file I came up with which works on Windows 10  In fact xgboost also caused issues for me. I don't remember excatly why, I think the build on conda for windows is simply outdated hence the solution is to go for a pip install.

Here my yml file for anyone interested which include mol2vec, rdkit, xgboost and more. (It's also usable from KNIME 4.3 for those who care). ipykernel is only needed if you want this environment to be available as a kernel in Juypter Notebook/lab, pywin32 probably also not needed if not on windows pc:

name: knime_py38
channels:  
  - defaults  
  - conda-forge  
  - plotly
dependencies:
  - python=3.8
  - pip  
  - matplotlib
  - numpy
  - pandas
  - pywin32
  - scikit-learn
  - scipy
  - seaborn
  - rdkit>=2020.09.02
  - tqdm
  - git
  - plotly
  - docopt
  - ipykernel
  - pip:
    - pyfancy
    - pyarrow >=1
    - gensim
    - xgboost
    - git+https://github.com/samoturk/mol2vec
